File: leetcode/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        match = {
            ")": "(",
            "]": "[",
            "}": "{",
        }

        for ch in s:
            if ch in match:
                if not stack or match.get(ch) != stack.pop():
                    return False
            else:
                stack.append(ch)

        return not stack

File: leetcode/0020-valid-parentheses/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = vec![];

        for b in s.bytes() {
            match b {
                b'(' | b'[' | b'{' => stack.push(b),
                b')' if matches!(stack.pop(), Some(b'(')) => continue,
                b']' if matches!(stack.pop(), Some(b'[')) => continue,
                b'}' if matches!(stack.pop(), Some(b'{')) => continue,
                _ => return false,
            }
        }

        stack.is_empty()
    }
}


